src/test/ui/impl-trait/impl_fn_associativity.rs

```diff
@@ -0,0 +1,16 @@
+// run-pass
+use std::fmt::Debug;
+
+fn f_debug() -> impl Fn() -> impl Debug {
+    || ()
+}
+
+fn ff_debug() -> impl Fn() -> impl Fn() -> impl Debug {
```

Comment:
Let's add some tests about `+` bounds:

```rust
// This should error because the returned closure is capturing
// its argument, which was not declared as part of the signature.
fn f_debug() -> impl Fn(&u8) -> impl Debug {
    |x| x
}
```

```rust
// This should (maybe) error because or parsing ambiguities
fn f_debug() -> impl Fn(&u8) -> impl Debug + '_ {
    |x| x
}
```


```rust
// This should (maybe) error because or parsing ambiguities
fn f_debug() -> impl Fn() -> impl Debug + Send {
    || ()
}
```

```rust
// This should work
fn f_debug() -> impl Fn(&u8) -> (impl Debug + '_) {
    |x| x
}
```


```rust
// This should work
fn f_debug() -> impl Fn() -> (impl Debug + Send) {
    || ()
}
```